### ISSUE TITLE
Add DSL.extractLarge() and DSL.epochLarge() to return Field<Long>

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#7795]
With the support of new, non-standard `DatePart` types, the result of the `DSL.extract()` method is no longer sufficient. `Field<Integer>` fails when extracting epochs in two ways:

- The last timestamp that can be represented by a 31 bit epoch is `2038-01-19 03:14:07`
- Epochs should be fractions in order to be able to represent fractional seconds

This is an incompatible change. The workaround is to use plain SQL to create a custom `EXTRACT(EPOCH FROM ...)` function

**Comment by yruchin** [comment]
Just wanted to express interest here as a user of jOOQ Professional: we've just run into exactly the first part of the issue (overflow aka [Y2038](https://en.wikipedia.org/wiki/Year_2038_problem)) in our project.

Is there any chance for the overflow to be fixed any time soon? It seems to be just a matter of widening the `Extract`'s function return type from `INT` to `BIGINT`. As such, can the partial fix be considered non-breaking and become a part of a 3.x release?

Of course, adding fractional part is more involved and is breaking, so I'm leaving it out of consideration.

**Comment by lukaseder** [comment]
> Is there any chance for the overflow to be fixed any time soon?

We don't like to implement breaking changes like this lightheartedly, or in isolation. There are numerous functions that could profit from changing their `Field<T>` type to some other `Field<T">` type, e.g. `count()`, which is a `Field<Integer>`, but should really be a `Field<Long>`.

> As such, can the partial fix be considered non-breaking and become a part of a 3.x release?

You need to extend this line of thought to that of a product with many users and customers. You think this is non-breaking, because it's important to *you*. Others think a similar thing is non-breaking because it's important to *them*. But the Venn diagram of "important" not-really-that-much-breaking changes doesn't have such a huge intersection area, as users tend to not agree on importance. You wouldn't like to have 10 other such not-strictly-breaking changes requested by other customers in the release that you're requesting here.

All the while there's a very simple workaround:

- Create your own `EXTRACT` function, e.g. using plain SQL templating: https://www.jooq.org/doc/latest/manual/sql-building/plain-sql-templating/
- Forbid usage of the built in function, e.g. using error prone: https://errorprone.info or some runtime agent that throws exceptions when the function is called

**Comment by oliveryasuna** [comment]
Looking forward to this.

**Comment by lukaseder** [comment]
Thanks for the ping, @oliveryasuna.

jOOQ 3.21 already has a new notion of "large" for certain methods / API calls, e.g. `executeLarge()`:

- https://github.com/jOOQ/jOOQ/issues/19261

Or `countLarge()`:

- https://github.com/jOOQ/jOOQ/issues/3714

So, `extractLarge()` would be the logical conclusion here. Fractional second support (e.g. when using `DatePart.EPOCH` or `DatePart.SECOND`) is the main reason why a `BigDecimal` seems preferrable to a `Long`. So the question is, should "large" really continue to mean `Long` (as with JDBC update counts, or `count()` calls), or can it extend to large fractional numbers?

I.e.:

- Add both `extractLarge():Field<Long>` and `extractDecimal():Field<BigDecimal>` (or similar name)
  - Alternatively, allow users to pass in the data type.
- Add only `extractLarge():Field<BigDecimal>`

**Comment by lukaseder** [comment]
Of course, any shortcomings can be worked around using plain SQL templates, as always.

**Comment by lukaseder** [comment]
ISO/IEC 9075-2:2023(E) 6.31 &lt;numeric value function> SR 7) specifies that the result type of the `<extract expression>` depends on the input values, where the `SECOND` `<extract field>` allows for maintaining the fractional precision in the expression's output. As such, the PostgreSQL behaviour is standards compliant, and jOOQ's API does not take this into account yet.

However, I still prefer the `Large` suffix to be consistent across API, meaning `Long`, so, I'd like to have more "overloads."

**Comment by lukaseder** [comment]
Of course, there's always also `extract(a, b).coerce(dataType)`, too!

**Comment by lukaseder** [comment]
Related: `epoch()` vs `epochLarge()`. Unlike `extract()`, which has many uses and thus both `extract()` and `extractLarge()` are viable, with `epoch()`, the current `Field<Integer>` value is really flawed.

I don't really like breaking compatibility for these things, but given that in "only" 13 years, `epoch()` becomes unusable, it's definitely a deprecation candidate at least:

- https://github.com/jOOQ/jOOQ/issues/19366

**Comment by lukaseder** [comment]
While this change can be implemented compatibly in the DSL API, there's still an incompatible change to the `QOM.Extract` API, which will now be generic: `Extract<T> <: Field<T>` instead of `Extract <: Field<Integer>`

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.21.
